File: backend/src/observability/logging.ts
import * as winston from 'winston';
import { getCurrentTraceId, getCurrentSpanId } from './tracing';
import { otelConfig } from '../config/otel.config';

const correlationFormat = winston.format((info) => {
  if (otelConfig.logging.correlationEnabled) {
    const traceId = getCurrentTraceId();
    const spanId = getCurrentSpanId();

    if (traceId) info.trace_id = traceId;
    if (spanId) info.span_id = spanId;
  }

  return info;
});

export const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || 'info',
  format: winston.format.combine(
    winston.format.timestamp(),
    correlationFormat(),
    winston.format.errors({ stack: true }),
    winston.format.json(),
  ),
  defaultMeta: {
    service: otelConfig.serviceName,
    version: otelConfig.serviceVersion,
    environment: otelConfig.environment,
  },
  transports: [
    new winston.transports.Console({
      format: winston.format.combine(
        winston.format.colorize(),
        winston.format.printf(({ level, message, timestamp, trace_id, span_id, ...meta }) => {
          const traceInfo = trace_id
            ? ` [trace:${trace_id.slice(0, 8)}${span_id ? ` span:${span_id.slice(0, 8)}` : ''}]`
            : '';
          const metaStr = Object.keys(meta).length ? ` ${JSON.stringify(meta)}` : '';
          return `${timestamp} ${level}${traceInfo}: ${message}${metaStr}`;
        }),
      ),
    }),
  ],
});

if (process.env.NODE_ENV === 'production') {
  logger.add(
    new winston.transports.File({
      filename: 'logs/error.log',
      level: 'error',
      maxsize: 10485760,
      maxFiles: 5,
    }),
  );

  logger.add(
    new winston.transports.File({
      filename: 'logs/combined.log',
      maxsize: 10485760,
      maxFiles: 10,
    }),
  );
}

export function createChildLogger(context: Record<string, any>) {
  return logger.child(context);
}

File: vision-service/src/observability/logging.py
import os
import logging
import json
from typing import Optional
from .tracing import get_current_trace_id, get_current_span_id

class CorrelatedJSONFormatter(logging.Formatter):
    """JSON formatter that includes trace correlation IDs."""
    
    def __init__(self):
        super().__init__()
        self.service_name = os.getenv("SERVICE_NAME", "smartpick-vision")
        self.environment = os.getenv("ENVIRONMENT", "development")
    
    def format(self, record: logging.LogRecord) -> str:
        log_data = {
            "timestamp": self.formatTime(record, self.datefmt),
            "level": record.levelname,
            "logger": record.name,
            "message": record.getMessage(),
            "service": self.service_name,
            "environment": self.environment,
        }
        
        # Add trace context if available
        trace_id = get_current_trace_id()
        span_id = get_current_span_id()
        
        if trace_id:
            log_data["trace_id"] = trace_id
        if span_id:
            log_data["span_id"] = span_id
        
        # Add exception info if present
        if record.exc_info:
            log_data["exception"] = self.formatException(record.exc_info)
        
        # Add any extra fields
        if hasattr(record, "extra"):
            log_data.update(record.extra)
        
        return json.dumps(log_data)


def setup_logging(level: str = "INFO") -> logging.Logger:
    """Setup application logging with correlation IDs.
    
    Args:
        level: Logging level (DEBUG, INFO, WARNING, ERROR, CRITICAL)
        
    Returns:
        Configured logger instance
    """
    log_level = os.getenv("LOG_LEVEL", level).upper()
    
    logger = logging.getLogger("smartpick-vision")
    logger.setLevel(log_level)
    
    # Remove existing handlers
    logger.handlers.clear()
    
    # Console handler
    console_handler = logging.StreamHandler()
    console_handler.setLevel(log_level)
    
    if os.getenv("ENVIRONMENT") == "production":
        # Use JSON formatting in production
        console_handler.setFormatter(CorrelatedJSONFormatter())
    else:
        # Use simpler formatting in development
        formatter = logging.Formatter(
            "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
        )
        console_handler.setFormatter(formatter)
    
    logger.addHandler(console_handler)
    
    return logger


def get_logger(name: Optional[str] = None) -> logging.Logger:
    """Get a logger instance with optional name.
    
    Args:
        name: Optional logger name, defaults to module name
        
    Returns:
        Logger instance
    """
    if name:
        return logging.getLogger(f"smartpick-vision.{name}")
    return logging.getLogger("smartpick-vision")
